lib/Conversion/PipelineToHW/PipelineToHW.cpp

```diff
@@ -367,6 +448,24 @@ class PipelineOutlineLowering : public PipelineLowering {
 
   // Handle to the instantiation of the last stage in the pipeline.
   hw::InstanceOp lastStageInst;
+
+  // A mapping between stages and the external inputs which they reference.
+  // A SetVector is used to ensure determinism in the order of the external
+  // inputs to a stage.
+  llvm::DenseMap<Block *, llvm::SetVector<Value>> stageExtInputs;
+
+  // A mapping between external inputs and their corresponding top-level
+  // input in the pipeline module.
+  llvm::DenseMap<Value, Value> toplevelExtInputs;
+
+  // Wrapper around stageExtInputs which returns a llvm::SmallVector<Value>,
+  // which in turn can be used to get a ValueRange (and by extension,
+  // TypeRange).
+  llvm::SmallVector<Value> getStageExtInputs(Block *stage) {
```

Comment:
I've seen you return a SmallVector a few times in your code... The more efficient way and (as a result) canonical to do this is with a 'populate vector' pattern. I'm OK with returning it so long as the vector is typically not long, but in this case we _could_ have many ext inputs.